ml/ml.cc

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "Config.h"
+#include "Dimension.h"
+#include "Host.h"
+
+using namespace ml;
+
+/*
+ * Assumptions:
+ *  1) hosts outlive their sets, and sets outlive their dimensions,
+ *  2) dimensions always have a set that has a host.
+ */
+
+void ml_init(void) {
+    Cfg.readMLConfig();
+}
+
+void ml_new_host(RRDHOST *RH) {
+    if (simple_pattern_matches(Cfg.SP_HostsToSkip, RH->hostname))
+        return;
+
+    Host *H = new Host(RH);
+    RH->ml_host = static_cast<ml_host_t>(H);
+
+    H->startAnomalyDetectionThreads();
+}
+
+void ml_delete_host(RRDHOST *RH) {
+    Host *H = static_cast<Host *>(RH->ml_host);
+    if (!H)
+        return;
+
+    H->stopAnomalyDetectionThreads();
+
+    delete H;
+    RH->ml_host = nullptr;
+}
+
+void ml_new_dimension(RRDDIM *RD) {
```

Comment:
TODO:

1. Sync with @andrewm4894  to support dims with `update-every != 1`
2. Hard-code the skipping of ML-related dimensions.